File: Day2/Problem_4.cpp
/* Data Structure :- Arrays(1-D & 2-D Arrays)
Problem Name :- Sort an array of 0s, 1s and 2s
Problem Statement :- Given an array nums consisting of only 0, 1, or 2. Sort the array in non-decreasing order. The sorting must be done in-place, without making a copy of the original array."""
*/

// TC : O(n) and SC : 0(1)

#include<bits/stdc++.h>
using namespace std;

void Sort_zeros_ones_twos(vector<int>& arr) { // Dutch National Flag Algorithm
    int n = arr.size();
    int low = 0;
    int mid = 0;
    int high = n - 1;

    while (mid <= high) {
        if (arr[mid] == 0) {
            swap(arr[low], arr[mid]);
            low++;
            mid++;
        }
        else if (arr[mid] == 1) {
            mid++;
        }
        else {
            swap(arr[mid], arr[high]);
            high--;
        }
    }
}

int main() {
    int n;
    cin >> n;

    vector<int> arr(n);

    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    Sort_zeros_ones_twos(arr);

    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }

    return 0;
}

File: Day2/Problem_4.py
""" Data Structure :- Arrays(1-D & 2-D Arrays)
Problem Name :- Sort an array of 0s, 1s and 2s
Problem Statement :- Given an array nums consisting of only 0, 1, or 2. Sort the array in non-decreasing order. The sorting must be done in-place, without making a copy of the original array.

(Time Complexity :- O(n) and Space Complexity :- O(1))

"""




def sort_zeros_ones_twos(nums):
    low, mid, high = 0, 0, len(nums) - 1

    while mid <= high:
        if nums[mid] == 0:
            nums[low], nums[mid] = nums[mid], nums[low]
            low += 1
            mid += 1

        elif nums[mid] == 1:
            mid += 1

        else:  # nums[mid] == 2
            nums[mid], nums[high] = nums[high], nums[mid]
            high -= 1


if __name__ == "__main__":
    n = int(input().strip())
    nums = list(map(int, input().split()))

    sort_zeros_ones_twos(nums)

    print(*nums)